Solve missing interpolation method (cubicspline)

khyox opened the pull request:
By commit 8bb2cc1 scipy.interpolate.CubicSpline method is
referenced in the pandas documentation (see pandas/core/generic.py)
but it is not wrapped by any interpolation method. This PR solves
this by adding the corresponding wrapper. SciPy's CubicSpline
is a cubic spline data interpolator that allows explicit control
of the boundary conditions for the interval.

Changes to be committed:
	modified:   pandas/core/missing.py
	modified:   pandas/tests/series/methods/test_interpolate.py

- [x] closes #xxxx (no issue opened but PR #20270 commented instead)
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

khyox committed:
Solve missing interpolation method (cubicspline)

By commit 8bb2cc1 scipy.interpolate.CubicSpline method is
referenced in the pandas documentation (see pandas/core/generic.py)
but it is not wrapped by any interpolation method. This commit
solves this adding the corresponding wrapper. SciPy's CubicSpline
is a cubic spline data interpolator that allows explicit control
of the boundary conditions for the interval.

Changes to be committed:
	modified:   ../../../core/missing.py
	modified:   test_interpolate.py

mroeschke commented on pandas/core/missing.py:
```diff
@@ -349,6 +351,13 @@ def _interpolate_scipy_wrapper(
     elif method == "akima":
         alt_methods["akima"] = _akima_interpolate
+    elif method == "cubicspline":
+        try:
```
I don't believe this try except is necessary since CubicSpline was introduced in 0.18.0 and our minimum supported scipy version is 0.19

khyox replied:
OK, I agree, thanks! Using the same argument, I think that another PR could address the removal of the try-except block for the PCHIP interpolator above.

mroeschke replied:
Yup that should be okay to remove if that was introduced before 0.19.0

khyox replied:
OK, thanks, I will remove both then.

mroeschke commented on pandas/core/missing.py:
```diff
@@ -478,6 +486,85 @@ def _akima_interpolate(xi, yi, x, der=0, axis=0):
+    """
+    from scipy import interpolate
+
+    P = interpolate.CubicSpline(
```
Could this just be directly assigned to `alt_methods`? Something similar to

`alt_methods["cubicspline"] = interpolate.CubicSpline(x, y, **kwargs)`

khyox replied:
Thanks. Like for `interpolate.Akima1DInterpolator`, `scipy.interpolate.CubicHermiteSpline` returns a `scipy.interpolate.PPoly` instance so, if further derivatives are needed now or in the future (like in `_akima_interpolate` or `_from_derivatives` methods), the full method could be a better approach. If so, with adding support for extracting derivatives for compatibility with those methods, it would be:
```
    if der == 0:
        return P(x)
    elif interpolate._isscalar(der):
        return P(x, der=der)
    else:
        return [P(x, nu) for nu in der]
```
as in `_akima_interpolate`, instead of just returning the evaluated polynomial `P(x)`. Any preference?

mroeschke replied:
Ah I see. Okay, how you have it looks good for now if we want to extend to support derivatives in the future

khyox replied:
OK, thanks!

jreback commented on pandas/core/missing.py:
```diff
@@ -478,6 +486,85 @@ def _akima_interpolate(xi, yi, x, der=0, axis=0):
         return [P(x, nu) for nu in der]
 
 
+def _cubicspline_interpolate(xi, yi, x, axis=0, bc_type="not-a-knot", extrapolate=None):
```
i don’t think we put a doc string for others
we should either link them instead

khyox replied:
Do you mean to remove the parameters of the docstring here that are already present in the SciPy method? I am following the previous logic of the methods `_from_derivatives` and `_akima_interpolate`. I agree that the doc for the `bc_type` parameter is quite verbose and it could be substituted by a link to the SciPy method doc. However, that same parameter (the control of boundary conditions) is what makes this function different from the simpler alternative using `scipy.interpolate.interp1d `with cubic splines and it could be worth having it explicitly in the docs. What do you think? Thanks.

jreback replied:
do we have doc strings for other interpolate methods that are just direct scipy calls?

khyox replied:
No, but as with `_from_derivatives` and `_akima_interpolate`, this is not just a direct SciPy call (see resolved conversation below with @mroeschke for details), that's why I added the doc string too.

khyox committed:
Remove try/except blocks (remnant of old SciPy releases support)

Changes to be committed:
	modified:   pandas/core/missing.py

khyox committed:
Add whatsnew entry and update the docs

Changes to be committed:
	modified:   doc/source/whatsnew/v1.1.0.rst
	modified:   pandas/core/generic.py

mroeschke commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -99,6 +99,7 @@ Other enhancements
   ``df.to_csv(path, compression={'method': 'gzip', 'compresslevel': 1}``
   (:issue:`33196`)
 - :meth:`~pandas.core.groupby.GroupBy.transform` has gained ``engine`` and ``engine_kwargs`` arguments that supports executing functions with ``Numba`` (:issue:`32854`)
+- :meth:`~pandas.core.resample.Resampler.interpolate` now supports SciPy interpolation method :class:`scipy.interpolate.CubicSpline` as method ``cubicspline``
```
Could you add the PR number to the end of this? `(:issue:PR number)`

khyox replied:
Sure, done by commit 7213e49. Thanks for the guidance.

mroeschke commented on pandas/tests/series/methods/test_interpolate.py:
```diff
@@ -97,6 +99,22 @@ def test_interpolate_time_raises_for_non_timeseries(self):
+        new_index = ser.index.union(Index([1.25, 1.5, 1.75, 2.25, 2.5, 2.75])).astype(
+            float
+        )
+        interp_s = ser.reindex(new_index).interpolate(method="cubicspline")
```
Could you rename this `result` and do the _getitem_ here so the next line could be `tm.assert_series.equal(result, expected)`?

khyox replied:
Done, thanks!

khyox committed:
Add PR number to the whatsnew entry

On branch CubicSpline
	modified:   doc/source/whatsnew/v1.1.0.rst

khyox committed:
Test: Clean final assert expression

On branch CubicSpline
	modified:   pandas/tests/series/methods/test_interpolate.py

khyox committed:
Restore ignore types for mypy ver 0.730 used in CI

Not needed for mypy version 0.740

jreback commented:
thanks @khyox very nice

khyox commented:
Many thanks @jreback and @mroeschke for your guidance on my 1st PR to the awesome pandas package!
